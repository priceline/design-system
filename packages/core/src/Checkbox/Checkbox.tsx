import { BoxChecked, BoxEmpty, BoxMinus } from 'pcln-icons'
import React, { type ComponentPropsWithRef, useEffect, useState } from 'react'
import styled from 'styled-components'
import { applyVariations, getPaletteColor } from '../utils/utils'

const CheckBoxWrapper = styled.div<CheckboxProps>`
  display: inline-flex;
  align-items: center;
  position: relative;
  vertical-align: middle;
  padding: 2px;
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  background-color: inherit;
  color: ${(props) =>
    props.disabled ? getPaletteColor('border.base')(props) : getPaletteColor('border.dark')(props)};

  svg {
    border: 1px solid transparent;
    border-radius: 4px;
    padding: 2px;
  }

  svg[data-name='checked'] {
    display: none;
  }
  svg[data-name='indeterminate'] {
    display: none;
  }

  ~ span {
    cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  }

  > input[data-indeterminate='true'] {
    & ~ svg[data-name='indeterminate'] {
      display: inline-block;
      color: ${(props) =>
        props.disabled ? getPaletteColor('border.base')(props) : getPaletteColor('base')(props)};
    }
    &:hover ~ svg[data-name='indeterminate'] {
      color: ${(props) =>
        props.disabled ? getPaletteColor('border.base')(props) : getPaletteColor('dark')(props)};
    }
    & ~ svg[data-name='empty'] {
      display: none;
    }
    & ~ svg[data-name='checked'] {
      display: none;
    }
  }

  > input:not([data-indeterminate='true']) {
    & ~ svg[data-name='empty'] {
      color: ${(props) =>
        props.disabled ? getPaletteColor('border.base') : getPaletteColor(props.unselectedColor)};
    }
    &:focus-visible ~ svg {
      border: 1px solid ${getPaletteColor('border.base')};
      background-color: ${getPaletteColor('background.light')};
    }
  }

  > input:checked:not([data-indeterminate='true']) {
    & ~ svg[data-name='checked'] {
      display: inline-block;
      color: ${(props) =>
        props.disabled ? getPaletteColor('border.base')(props) : getPaletteColor('base')(props)};
    }

    & ~ svg[data-name='empty'] {
      display: none;
    }

    &:focus-visible ~ svg {
      border: 1px solid ${getPaletteColor('base')};
      background-color: ${getPaletteColor('light')};
    }

    &:hover ~ svg[data-name='checked'] {
      color: ${(props) =>
        props.disabled ? getPaletteColor('border.base')(props) : getPaletteColor('dark')(props)};
    }
  }

  > input:hover ~ svg[data-name='empty'] {
    color: ${(props) =>
      props.disabled ? getPaletteColor('border.base')(props) : getPaletteColor('base')(props)};
  }

  ${applyVariations('Checkbox')}
`

const StyledInput = styled.input`
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
`

/**
 * @public
 */
export type CheckboxProps = ComponentPropsWithRef<'input'> & {
  id?: string
  indeterminate?: boolean
  size?: number
  color?: string
  checked?: boolean
  defaultChecked?: boolean
  disabled?: boolean
  unselectedColor?: string
}

/**
 * @public
 */
export const Checkbox = React.forwardRef<HTMLInputElement, CheckboxProps>(
  (
    {
      size = 20,
      indeterminate = false,
      color = 'primary',
      unselectedColor = 'text.light',
      disabled,
      onChange,
      defaultChecked,
      ...props
    },
    ref
  ) => {
    const spreadProps = {
      size,
      indeterminate,
      color,
      unselectedColor,
      disabled,
      onChange,
      defaultChecked,
      ...props,
    }
    const [showIndeterminate, setShowIndeterminate] = useState(indeterminate && !defaultChecked)

    function handleChange(e) {
      setShowIndeterminate(false)
      if (typeof onChange === 'function') {
        onChange(e)
      }
    }

    useEffect(() => {
      setShowIndeterminate(indeterminate)
    }, [indeterminate])

    // Add 4px to Icon's height and width to account for size reduction caused by adding padding to SVG element
    const borderAdjustedSize = size + 4
    return (
      <CheckBoxWrapper {...spreadProps}>
        <StyledInput
          type='checkbox'
          {...spreadProps}
          onChange={handleChange}
          aria-checked={showIndeterminate ? 'mixed' : props.checked}
          data-indeterminate={showIndeterminate}
          ref={ref}
        />
        <BoxChecked size={borderAdjustedSize} data-name='checked' />
        <BoxMinus size={borderAdjustedSize} data-name='indeterminate' />
        <BoxEmpty size={borderAdjustedSize} data-name='empty' />
      </CheckBoxWrapper>
    )
  }
)

Checkbox.displayName = 'Checkbox'
