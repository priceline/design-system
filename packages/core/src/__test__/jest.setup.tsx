import '@testing-library/jest-dom'
import 'jest-styled-components'

global.requestAnimationFrame =
  global.requestAnimationFrame ||
  function _raf(cb) {
    return setTimeout(cb, 0)
  }
