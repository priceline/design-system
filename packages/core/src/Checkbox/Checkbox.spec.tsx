import React from 'react'
import { fireEvent, render } from '../__test__/testing-library'
import { Checkbox } from './Checkbox'

describe('Checkbox', () => {
  const onChange = jest.fn()

  test('renders without the theme passed specifically', () => {
    const { asFragment } = render(<Checkbox id='check-box' onChange={onChange} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders checked when defaultChecked prop is passed as true', () => {
    const { container } = render(<Checkbox id='check-box' defaultChecked onChange={onChange} />)
    const checkbox = container.querySelector('[type=checkbox]') as HTMLInputElement
    expect(checkbox.checked).toBe(true)
  })

  test('renders disabled with disabled prop', () => {
    const { asFragment } = render(<Checkbox id='check-box' disabled onChange={onChange} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('renders disabled with defaultChecked', () => {
    const { asFragment } = render(
      <Checkbox id='check-box' disabled={true} defaultChecked={true} onChange={onChange} />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  test('calls onChange when clicked', () => {
    const { container } = render(<Checkbox id='check-box' onChange={onChange} />)
    const checkbox = container.querySelector('[type=checkbox]') as HTMLInputElement
    fireEvent.click(checkbox)
    expect(onChange).toHaveBeenCalled()
  })

  it('renders an indeterminate checkbox that can be clicked to set checked to true', () => {
    const { container } = render(<Checkbox id='check-box' indeterminate onChange={onChange} />)
    const checkbox = container.querySelector('[type=checkbox]') as HTMLInputElement

    expect(checkbox.checked).toBe(false)
    fireEvent.click(checkbox)
    expect(checkbox.checked).toBe(true)
  })
  it('renders an indeterminate checkbox that can be clicked to set checked to false', () => {
    const { container } = render(<Checkbox id='check-box' indeterminate defaultChecked onChange={onChange} />)
    const checkbox = container.querySelector('[type=checkbox]') as HTMLInputElement

    expect(checkbox.checked).toBe(true)
    // expect(checkbox.indeterminate).toBe(true)
    fireEvent.click(checkbox)
    expect(checkbox.checked).toBe(false)
    // expect(checkbox.indeterminate).toBe(false)
  })
  it('correctly passes in the ref so that the underlying input element can be modified by the parent component if needed', () => {
    const ref = React.createRef()
    const { getByRole } = render(
      // @ts-ignore
      <Checkbox id='check-box' ref={ref} indeterminate defaultChecked onChange={onChange} />
    )
    const checkbox = getByRole('checkbox') as HTMLInputElement

    // @ts-ignore
    expect(ref.current.id).toBe(checkbox.id)
  })
})
