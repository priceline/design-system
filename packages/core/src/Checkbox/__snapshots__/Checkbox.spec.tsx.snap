// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders disabled with defaultChecked 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  line-height: 1;
  width: 24px;
}

.c4 {
  outline: none;
}

.c6 {
  outline: none;
}

.c5 {
  outline: none;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  vertical-align: middle;
  padding: 2px;
  cursor: not-allowed;
  background-color: inherit;
  color: #c0cad5;
}

.c1 svg {
  border: 1px solid transparent;
  border-radius: 4px;
  padding: 2px;
}

.c1 svg[data-name='checked'] {
  display: none;
}

.c1 svg[data-name='indeterminate'] {
  display: none;
}

.c1 ~ span {
  cursor: not-allowed;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='indeterminate'] {
  display: inline-block;
  color: #c0cad5;
}

.c1 > input[data-indeterminate='true']:hover ~ svg[data-name='indeterminate'] {
  color: #c0cad5;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='empty'] {
  display: none;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='checked'] {
  display: none;
}

.c1 > input:not([data-indeterminate='true']) ~ svg[data-name='empty'] {
  color: #c0cad5;
}

.c1 > input:not([data-indeterminate='true']):focus-visible ~ svg {
  border: 1px solid #c0cad5;
  background-color: #f4f6f8;
}

.c1 > input:checked:not([data-indeterminate='true']) ~ svg[data-name='checked'] {
  display: inline-block;
  color: #c0cad5;
}

.c1 > input:checked:not([data-indeterminate='true']) ~ svg[data-name='empty'] {
  display: none;
}

.c1 > input:checked:not([data-indeterminate='true']):focus-visible ~ svg {
  border: 1px solid #0068ef;
  background-color: #e8f2ff;
}

.c1 > input:checked:not([data-indeterminate='true']):hover ~ svg[data-name='checked'] {
  color: #c0cad5;
}

.c1 > input:hover ~ svg[data-name='empty'] {
  color: #c0cad5;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
}

<div
    class="c0"
  >
    <div
      class="c1"
      color="primary"
      disabled=""
    >
      <input
        checked=""
        class="c2"
        color="primary"
        data-indeterminate="false"
        disabled=""
        id="check-box"
        role="checkbox"
        size="20"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="c3 c4"
        data-name="checked"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424-424l-86-86q-11-11-28-11t-28 11q-11 11-11 28t11 28l114 114q12 12 28 12t28-12l226-226q11-11 11-28t-11-28q-11-11-28-11t-28 11L424-424zM200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c3 c5"
        data-name="indeterminate"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M320-440h320q17 0 28.5-11.5T680-480q0-17-11.5-28.5T640-520H320q-17 0-28.5 11.5T280-480q0 17 11.5 28.5T320-440zM200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200zm0-80h560v-560H200v560zm0-560v560-560z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c3 c6"
        data-name="empty"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200zm0-80h560v-560H200v560z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Checkbox renders disabled with disabled prop 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  line-height: 1;
  width: 24px;
}

.c4 {
  outline: none;
}

.c6 {
  outline: none;
}

.c5 {
  outline: none;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  vertical-align: middle;
  padding: 2px;
  cursor: not-allowed;
  background-color: inherit;
  color: #c0cad5;
}

.c1 svg {
  border: 1px solid transparent;
  border-radius: 4px;
  padding: 2px;
}

.c1 svg[data-name='checked'] {
  display: none;
}

.c1 svg[data-name='indeterminate'] {
  display: none;
}

.c1 ~ span {
  cursor: not-allowed;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='indeterminate'] {
  display: inline-block;
  color: #c0cad5;
}

.c1 > input[data-indeterminate='true']:hover ~ svg[data-name='indeterminate'] {
  color: #c0cad5;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='empty'] {
  display: none;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='checked'] {
  display: none;
}

.c1 > input:not([data-indeterminate='true']) ~ svg[data-name='empty'] {
  color: #c0cad5;
}

.c1 > input:not([data-indeterminate='true']):focus-visible ~ svg {
  border: 1px solid #c0cad5;
  background-color: #f4f6f8;
}

.c1 > input:checked:not([data-indeterminate='true']) ~ svg[data-name='checked'] {
  display: inline-block;
  color: #c0cad5;
}

.c1 > input:checked:not([data-indeterminate='true']) ~ svg[data-name='empty'] {
  display: none;
}

.c1 > input:checked:not([data-indeterminate='true']):focus-visible ~ svg {
  border: 1px solid #0068ef;
  background-color: #e8f2ff;
}

.c1 > input:checked:not([data-indeterminate='true']):hover ~ svg[data-name='checked'] {
  color: #c0cad5;
}

.c1 > input:hover ~ svg[data-name='empty'] {
  color: #c0cad5;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
}

<div
    class="c0"
  >
    <div
      class="c1"
      color="primary"
      disabled=""
    >
      <input
        class="c2"
        color="primary"
        data-indeterminate="false"
        disabled=""
        id="check-box"
        role="checkbox"
        size="20"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="c3 c4"
        data-name="checked"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424-424l-86-86q-11-11-28-11t-28 11q-11 11-11 28t11 28l114 114q12 12 28 12t28-12l226-226q11-11 11-28t-11-28q-11-11-28-11t-28 11L424-424zM200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c3 c5"
        data-name="indeterminate"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M320-440h320q17 0 28.5-11.5T680-480q0-17-11.5-28.5T640-520H320q-17 0-28.5 11.5T280-480q0 17 11.5 28.5T320-440zM200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200zm0-80h560v-560H200v560zm0-560v560-560z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c3 c6"
        data-name="empty"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200zm0-80h560v-560H200v560z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Checkbox renders without the theme passed specifically 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Montserrat','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.4;
  font-weight: 500;
}

.c0 * {
  box-sizing: border-box;
}

.c3 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  line-height: 1;
  width: 24px;
}

.c4 {
  outline: none;
}

.c6 {
  outline: none;
}

.c5 {
  outline: none;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  vertical-align: middle;
  padding: 2px;
  cursor: pointer;
  background-color: inherit;
  color: #5e6872;
}

.c1 svg {
  border: 1px solid transparent;
  border-radius: 4px;
  padding: 2px;
}

.c1 svg[data-name='checked'] {
  display: none;
}

.c1 svg[data-name='indeterminate'] {
  display: none;
}

.c1 ~ span {
  cursor: pointer;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='indeterminate'] {
  display: inline-block;
  color: #0068ef;
}

.c1 > input[data-indeterminate='true']:hover ~ svg[data-name='indeterminate'] {
  color: #049;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='empty'] {
  display: none;
}

.c1 > input[data-indeterminate='true'] ~ svg[data-name='checked'] {
  display: none;
}

.c1 > input:not([data-indeterminate='true']) ~ svg[data-name='empty'] {
  color: #4f6f8f;
}

.c1 > input:not([data-indeterminate='true']):focus-visible ~ svg {
  border: 1px solid #c0cad5;
  background-color: #f4f6f8;
}

.c1 > input:checked:not([data-indeterminate='true']) ~ svg[data-name='checked'] {
  display: inline-block;
  color: #0068ef;
}

.c1 > input:checked:not([data-indeterminate='true']) ~ svg[data-name='empty'] {
  display: none;
}

.c1 > input:checked:not([data-indeterminate='true']):focus-visible ~ svg {
  border: 1px solid #0068ef;
  background-color: #e8f2ff;
}

.c1 > input:checked:not([data-indeterminate='true']):hover ~ svg[data-name='checked'] {
  color: #049;
}

.c1 > input:hover ~ svg[data-name='empty'] {
  color: #0068ef;
}

.c2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  opacity: 0;
  position: absolute;
  z-index: 0;
}

<div
    class="c0"
  >
    <div
      class="c1"
      color="primary"
    >
      <input
        class="c2"
        color="primary"
        data-indeterminate="false"
        id="check-box"
        role="checkbox"
        size="20"
        type="checkbox"
      />
      <svg
        aria-hidden="true"
        class="c3 c4"
        data-name="checked"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M424-424l-86-86q-11-11-28-11t-28 11q-11 11-11 28t11 28l114 114q12 12 28 12t28-12l226-226q11-11 11-28t-11-28q-11-11-28-11t-28 11L424-424zM200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c3 c5"
        data-name="indeterminate"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M320-440h320q17 0 28.5-11.5T680-480q0-17-11.5-28.5T640-520H320q-17 0-28.5 11.5T280-480q0 17 11.5 28.5T320-440zM200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200zm0-80h560v-560H200v560zm0-560v560-560z"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="c3 c6"
        data-name="empty"
        fill="currentcolor"
        focusable="false"
        height="24"
        role="img"
        tabindex="-1"
        viewBox="0 -960 960 960"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M200-120q-33 0-56.5-23.5T120-200v-560q0-33 23.5-56.5T200-840h560q33 0 56.5 23.5T840-760v560q0 33-23.5 56.5T760-120H200zm0-80h560v-560H200v560z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
